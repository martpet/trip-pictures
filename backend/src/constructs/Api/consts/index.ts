export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
