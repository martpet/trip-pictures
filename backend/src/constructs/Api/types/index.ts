export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
