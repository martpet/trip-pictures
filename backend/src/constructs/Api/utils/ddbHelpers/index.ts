export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
