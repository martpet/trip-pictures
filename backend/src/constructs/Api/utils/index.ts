export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
