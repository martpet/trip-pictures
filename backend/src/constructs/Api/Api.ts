import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
