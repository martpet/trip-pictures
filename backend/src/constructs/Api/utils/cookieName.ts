import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
