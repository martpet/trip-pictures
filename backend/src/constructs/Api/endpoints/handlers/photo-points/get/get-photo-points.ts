import {
  APIGatewayProxyHandlerV2,
  GetPhotoPointsResponse,
  PhotosTableItem,
  photosTableOptions,
} from 'lambda-layer';

import { ddbScan } from '~/constructs/Api/utils';

export const handler: APIGatewayProxyHandlerV2<GetPhotoPointsResponse> = async () => {
  const items = await ddbScan<Pick<PhotosTableItem, 'fingerprint'>>({
    tableOptions: photosTableOptions,
    attributes: ['fingerprint'],
  });

  return items.sort((a, b) => b.fingerprint.localeCompare(a.fingerprint));
};
