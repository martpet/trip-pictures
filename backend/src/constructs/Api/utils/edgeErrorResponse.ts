import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
