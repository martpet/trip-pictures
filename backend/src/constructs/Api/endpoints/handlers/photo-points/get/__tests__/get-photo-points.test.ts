import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { ddbScan, itCalls, itResolves } from '~/constructs/Api/utils';

import { handler } from '../get-photo-points';

vi.mock('~/constructs/Api/utils/ddbHelpers/ddbScan');

const ddbMock = mockClient(DynamoDBDocumentClient);

vi.mocked(ddbScan).mockResolvedValue([
  { fingerprint: 'c' },
  { fingerprint: 'a' },
  { fingerprint: 'b' },
]);

beforeEach(() => {
  ddbMock.reset();
  ddbMock.on(GetCommand).resolves({ Item: { dummyPhotoProp: 'dummyPhotoValue' } });
});

const args = [] as unknown as Parameters<typeof handler>;

describe('get-photo-points', () => {
  itCalls(ddbScan, handler, args);
  itResolves(handler, args);
});
