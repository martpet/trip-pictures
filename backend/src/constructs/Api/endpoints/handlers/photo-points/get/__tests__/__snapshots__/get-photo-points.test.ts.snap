// Vitest Snapshot v1

exports[`get-photo-points > calls "ddbScan" with correct args 1`] = `
[
  [
    {
      "attributes": [
        "fingerprint",
      ],
      "tableOptions": {
        "partitionKey": {
          "name": "fingerprint",
          "type": "S",
        },
        "stream": "NEW_IMAGE",
        "tableName": "TP-photos",
      },
    },
  ],
]
`;

exports[`get-photo-points > resolves with a correct value 1`] = `
[
  {
    "fingerprint": "c",
  },
  {
    "fingerprint": "b",
  },
  {
    "fingerprint": "a",
  },
]
`;
